### ISSUE TITLE
Video doesn't have event for loading/processing error

### ISSUE THREAD

**Issue by germn** [kivy/kivy#4953]
If we'll try to load broken video file, we will see error message in log (which is different for different providers), but looks like we just can't catch it somehow in code.

For there's no way example to tell user file is broken as soon as we couldn't load it.

This is probably related to error while playing loaded video.

**Comment by Zen-CODE** [comment]
Please could you post a concrete example? A problematic file may well be swallowed by searching for video providers, but a concrete example would help narrow this down.

**Comment by germn** [comment]
Example might be simplest as possible:
http://pastebin.com/5SiHrSeU

Files for test:
broken.avi - https://yadi.sk/d/qU5kjpN03Esx82
normal.avi - https://yadi.sk/i/au0NqZp43EsxA5

When we trying to load broken file, we got error in console (
> [VideoGstplayer] Internal data stream error.

or

> [ffpyplayer  ] [avi @ 00000000046cb280] Something went wrong during header parsing, I will ignore it and try to continue anyway.

) and no way to catch error event somehow (since `loaded` changed only for normal video[*]).

Both gstplayer and ffpyplayer providers handle errors with logging it and not somehow telling outer code about what happened:
https://github.com/kivy/kivy/blob/master/kivy/core/video/video_gstplayer.py#L41
https://github.com/kivy/kivy/blob/master/kivy/core/video/video_ffpyplayer.py#L75


[*] Btw `loaded` works only for first source, there's [fix](https://github.com/kivy/kivy/pull/4974) for it. This is unrelated to this issue.

**Comment by Zen-CODE** [comment]
I suppose an 'on_error' event would be the right way to handle it? There are so many available providers, though, it would not be a trivial task to add and test this with each.

@matham. A few questions
1. Would you agree an 'on_error' event is best?
2. Which providers would need to be covered in order to add the documented support for 'on_error'? All, or just the defaults?
3. The various video providers seem to be documented, but not part of the doc generation. Should they be added? You currently have to open the raw files to see them...

**Comment by germn** [comment]
> I suppose an 'on_error' event would be the right way to handle it?

I think that's what we need. We would be able to bind to `loaded` and `on_error` to handle both possible situations. 

> There are so many available providers, though, it would not be a trivial task to add and test this with each.

You talk about all providers, not only video providers, right? Because for video we need only `gstplayer` and `ffpyplayer` (`ffmpeg` is probably [outdated](https://groups.google.com/d/msg/kivy-users/ypJOZXlUEqM/aMToyDAY66UJ) and it seems there's no much sense to support new features for it).

While googling about other providers found [this](https://github.com/kivy/kivy/issues/2991). Looks like some other widges already have on_error event.

**Comment by Zen-CODE** [comment]
Well, there are LOTS of providers: input  providers, video provider, audio providers, text providers etc. So doing that for all is a nice idea but a huge effort. Just for video, there are:
```
    video_ffmpeg.py
    video_ffpyplayer.py
    video_gi.py
    video_gstplayer.py
    video_pyglet.py
    video_pygst.py
```
Some of these are deprecated, but you get the idea :-). Let me look into implementing it in one, and see how it goes. Which is most important to you? gstplayer?

**Comment by germn** [comment]
>  Just for video, there are

I might be wrong, but most of them were removed recently:
https://github.com/kivy/kivy/tree/master/kivy/core/video

> Which is most important to you? gstplayer?

Personally for me it's `ffpyplayer` since I'm creating app for Android. `gstplayer` should be most important provider for people who create apps for desktop.

**Comment by Zen-CODE** [comment]
Yes, that is what I meant by deprecated :-), But to the point. This is an area that I am not familiar with. Lets add that functionality to ffpyplayer, then extend it to the other providers once you are happy that it work properly for you?

**Comment by matham** [comment]
Yeah, a `on_error` seems fine. It probably should take the exception object and possibly `sys.exc_info()`. Not sure about the latter though.

**Comment by Julian-O** [comment]
There is a Has PR tag, but I don't see the PR.

@Zen-CODE: Did you ever get your prototype working?

**Comment by Zen-CODE** [comment]
@Julian-O   Not  with ffpyplayer. I kept getting compilation issues. What I ended up doing was writing a kivy soundplayer compatible  wrapper around python-vlc. This was really stable on Windows, Linux and Mac but as it needs a VLC install, was not really suitable for mobile. So I never submitted a PR for it for this reason. But it works damn well, I pretty much have this running all day every day without issues.....

https://github.com/Zen-CODE/zenplayer/tree/master/zenplayer/components/audio
